configuration_broker/config_broker/config_broker.h

```diff
@@ -0,0 +1,145 @@
+                                                                               \
+	DECLARE_AND_DEFINE_STATIC_SEALED_VALUE(                                    \
+	  struct {                                                                 \
+		  uint16_t   id;                                                       \
+		  size_t     notUsed1;                                                 \
+		  uint32_t   notUsed2;                                                 \
+		  const char Name[sizeof(name)];                                       \
+	  },                                                                       \
+	  config_broker,                                                           \
+	  WriteConfigKey,                                                          \
+	  __write_config_capability_##name,                                        \
+	  0,                                                                       \
+	  0,                                                                       \
+	  0,                                                                       \
+	  name);
+
+#define WRITE_CONFIG_CAPABILITY(name)                                          \
+	STATIC_SEALED_VALUE(__write_config_capability_##name)
+
+//
+// Create a Sealed Capability to set the parser for a config item
+//
+#define DEFINE_PARSER_CONFIG_CAPABILITY(name, Size, UpdateInterval)            \
+                                                                               \
+	DECLARE_AND_DEFINE_STATIC_SEALED_VALUE(                                    \
+	  struct {                                                                 \
+		  uint16_t   id;                                                       \
+		  size_t     size;                                                     \
+		  uint32_t   update_interval;                                          \
+		  const char Name[sizeof(name)];                                       \
+	  },                                                                       \
+	  config_broker,                                                           \
+	  ParserConfigKey,                                                         \
+	  __parser_config_capability_##name,                                       \
+	  0,                                                                       \
+	  Size,                                                                    \
+	  UpdateInterval,                                                          \
+	  name);
+
+#define PARSER_CONFIG_CAPABILITY(name)                                         \
+	STATIC_SEALED_VALUE(__parser_config_capability_##name)
+
+//
+// External view of a configuration item.
+//
+struct ConfigItem
+{
+	const char            *name;         // name
+	uint32_t               version;      // version
+	void                  *data;         // value
+	std::atomic<uint32_t> *versionFutex; // Futex to wait for version change
```

Comment:
Probably not important, but you could save a bit of memory if you moved `name`, `data`, and `versionFutex` into a separate structure and embedded this in `ConfigItemInternal` and just returned a single read-only capability to this.  This structure then becomes a pair of last-seen-version and the const pointer to the item information.

Reply:
My thinking in separating out the version value from the futex was to avoid that the consumer ever has to read the version value from there outside of the context of the wait.   Even as an atomic it seems like there was potential for a race condition in there between threads in different consumers if I leave it up to them when to read the version value.  With the code as it is each consumer gets their own stack object with a consistent version and data pointer, where although *data and *versionFutex are shared they are readonly.  Even if there is an update between this struct being returned and the called reading it  data and version are always consistent.